californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -234,6 +235,13 @@ public Exchange receiveResponse(final Response response, final CorrelationContex
 			return null;
 		} else if (correlationContextMatcher.isResponseRelatedToRequest(exchange.getCorrelationContext(), responseContext)) {
 
+			if (response.getType() == Type.ACK && exchange.getCurrentRequest().getMID() != response.getMID()) {
```

Comment:
I have a very hard time following and reviewing your PRs, @boaks. Most of the time, it is very hard for me to understand the context of the proposes change(s). More detail information in the description fields and within the code would help a lot.
Please add a (detailed) comment describing under which circumstances you expect this to happen and why it makes sense to return `null`.